core/dbt/context/providers.py

```diff
@@ -99,6 +100,62 @@ def type(self):
     def commit(self):
         return self.adapter.commit_if_has_connection()
 
+    def _get_adapter_macro_prefixes(self) -> List[str]:
+        # a future version of this could have plugins automatically call fall
+        # back to their dependencies' dependencies by using
+        # `get_adapter_type_names` instead of `[self.config.credentials.type]`
+        search_prefixes = [self.adapter.type(), 'default']
+        return search_prefixes
+
+    def dispatch(
+        self, macro_name: str, packages: Optional[List[str]] = None
+    ) -> MacroGenerator:
+        search_packages: List[Optional[str]]
+
+        if '.' in macro_name:
+            suggest_package, suggest_macro_name = macro_name.split('.', 1)
+            msg = (
+                f'In adapter.dispatch, got a macro name of "{macro_name}", '
+                f'but "." is not a valid macro name component. Did you mean '
+                f'`adapter.dispatch("{suggest_macro_name}", '
+                f'packages=["{suggest_package}"])`?'
+            )
+            raise CompilationException(msg)
+
+        if packages is None:
+            search_packages = [None]
+        else:
+            search_packages = packages
+
+        attempts = []
+
+        for package_name in search_packages:
+            for prefix in self._get_adapter_macro_prefixes():
```

Comment:
Am I right in understanding that the order of preference is `packages` > adapter specificity? So e.g. a dispatch with two packages running on Postgres would look for:
1. `packages[0]`, `postgres_`
2. `packages[0]`, `default__`
3. `packages[1]`, `postgres__`
4. `packages[1]`, `default__`

Per the comment above, someday we may make it so that running on Redshift looks for:
1. `packages[0]`, `redshift__`
2. `packages[0]`, `postgres__`
3. `packages[0]`, `default__`
4. `packages[1]`, `redshift__`
...

Reply:
Yes. We can of course swap this order if we want, but it seemed more correct to me to go by package first - looking at the example in #2302, it seems that you'd want your custom macro to tell dbt-utils "search this other package first and then fall back to dbt-utils if it's not there". This provides a nice escape hatch for issues where "dbt-utils has an implementation of a macro that doesn't work for me because of my obscure workflow". Or even just a bug!

Reply:
Agree! Give first priority to the place where end users have the most control, let them override `default__` or do anything they want. Just wanted to make sure I had this down before writing docs